filebeat/registrar/registrar.go

```diff
@@ -269,45 +115,116 @@ func (r *Registrar) Start() error {
 	}
 
 	r.wg.Add(1)
-	go r.Run()
+	go func() {
+		defer r.wg.Done()
+		r.Run()
+	}()
 
 	return nil
 }
 
+// Stop stops the registry. It waits until Run function finished.
+func (r *Registrar) Stop() {
+	logp.Info("Stopping Registrar")
+
+	close(r.done)
+	r.wg.Wait()
+	logp.Info("Registrar stopped")
+}
+
 func (r *Registrar) Run() {
 	logp.Debug("registrar", "Starting Registrar")
-	// Writes registry on shutdown
+	defer logp.Debug("registrar", "Stopping Registrar")
+
+	defer r.store.Close()
+
 	defer func() {
-		r.writeRegistry()
-		r.wg.Done()
+		writeStates(r.store, r.states.GetStates())
 	}()
 
 	var (
+		// we keep a long running write transaction.
+		// The 'filebeat' store must not be used outside of the Registrar.
 		timer  *time.Timer
 		flushC <-chan time.Time
+
+		directIn  chan []file.State
+		collectIn chan []file.State
 	)
 
+	if r.flushTimeout <= 0 {
+		directIn = r.Channel
+	} else {
+		collectIn = r.Channel
+	}
+
 	for {
 		select {
 		case <-r.done:
 			logp.Info("Ending Registrar")
 			return
-		case <-flushC:
-			flushC = nil
-			timer.Stop()
-			r.flushRegistry()
-		case states := <-r.Channel:
+
+		case states := <-directIn:
+			// no flush timeout configured. Directly update registry
+			r.onEvents(states)
+			if err := r.commitStateUpdates(); err != nil {
+				r.failing(err)
```

Comment:
What is the point of adding `failing` at the moment? The function `commitStateUpdate` does not return any errors.

Reply:
Also, it is does make sense to send anything to `r.done` from `failing` as in this case `Run` returns on error.

Reply:
It's and artifact from an older implementation I have had that would only write updated state to the statestore (reduce number of required updates and writes). Will remove it.

Although it would be nice, I found the logic to become quite 'complex' and removed it again in favor if the old logic of always writing all state updates (writing delta-updates only did clash with the gc provided by file.states).

In order to prevent the store to become inconsistent with state-updates I forced filebeat to block, preventing filebeat from sending any new events => this would force an operator to have a look and check why the registry can't be written (e.g. disk full, permissions changed, ...)

But reverting to the old logic we can just try to write the state on the next update. Worst case is the registry can never be written to and filebeat will send all logs again.